src/transformStateWithClones.js

```diff
@@ -7,7 +7,52 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const statesHistory = [];
+  let newState = state;
+
+  for (const action of actions) {
+    newState = { ...newState };
+
+    switch (action.type) {
+      case 'addProperties':
+        newState = addProperties(newState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        newState = removeProperties(newState, action.keysToRemove);
+        break;
+
+      case 'clear':
+        newState = {};
+        break;
+    }
+
+    statesHistory.push(newState);
```

Comment:
[CODE STYLE] - It is better to push a copy of the object to the array at the end of the each loop cycle, but outside of the `switch` block, after current action is processed. This ensures that the same copy of the state isn't mutated by subsequent actions.